Source/JavaScriptCore/wasm/WasmCallee.h

```diff
@@ -265,7 +265,7 @@ class LLIntCallee final : public Callee {
 
     LLIntTierUpCounter& tierUpCounter() { return m_tierUpCounter; }
 
-    const FunctionSignature& signature(unsigned index) const
+    const TypeDefinition& signature(unsigned index) const
```

Comment:
This change was intentional, and I think it's required. The reason is that with the addition of recursion groups, in generation type signatures have to be compared by projection, even if you know it will be a function signature underneath.

The reason is you may have function types like the following:

```
(rec (type (func $f1)) (type (struct ...)))

(rec (type (func $f2)))
```

Even though `$f1` and `$f2` look the same, and are indeed the same if you unfold the projections, they are not equal types because they are in different recursion groups.

(it seems a bit silly in this very simple case, but you can construct more interesting examples where `$f1` and `$f2` would unfold to the same type that, say, reference another member of the recursion group. And where these other members are not equal types.)

Reply:
@kmiller68 was trying to explain this to me, so I am wondering if I can ask few clarifying questions: 

```
(rec $r1
    (type $tree (struct (field i32) (field (ref $forest))))
    (type $forest (struct (field (ref $tree)) (field (ref $forest)))))
 
(type $s1 (struct (field (ref $tree))))
```

1) Types are stored as instances of TypeInformation, and expanded to produce a FunctionSignature (including for struct types)?

1) Projection in this case means taking a TypeDefinition, and selecting an element of it without any substitution? So:

$r1 is given a type index
$s1 is represented by TypeDefinition(StructType(Projection($r1, 0)))
$tree is TypeDefinition(Projection($r1, 0))

1) Expanding replaces a projection with a new projection that has had one step of unrolling applied, with the recursion group as the target of the unrolling

1) Unrolling is applying one step of substitution, replacing references to the recursion group with its expansion. So:

expand($tree) <=>
unroll(Projection($r1, 0) with respect to $r1) <=>
TypeDefinition(StructType(i32), TypeDefinition(Projection($r1, 0)))

This new expansion is given its own type index, and comparisons are made by type index?

We need to compare types by type index because the type definitions are iso-recursive? So a TypeDefinition and its .expand() are not the same type? 

If they were equirecursive, then we would have to perform a dfs expanding until we could prove that the types did/did not match?

Sorry to bug you with all these questions, I will continue playing with this patch to see how many of these questions I can answer by myself.